Iter8

luckyseadog opened the pull request.

luckyseadog committed:
json API

luckyseadog committed:
in server: value/ - > value and update/ -> update

luckyseadog committed:
added old methods

luckyseadog committed:
regexp

luckyseadog committed:
regexp

luckyseadog committed:
2 minutes of running agent

luckyseadog committed:
update - > update/

luckyseadog committed:
fix

luckyseadog committed:
continue instead return in agent

luckyseadog committed:
if in VALUE req we get array of len 1 than we retun json that looks like {}

luckyseadog committed:
fix

luckyseadog committed:
Additional changes after resolving merge conflicts

luckyseadog committed:
starage

luckyseadog committed:
Add support of ADDRESS POLL_INTERVAL REPORT_INTERVAL system variables
if error is occured then log.Fatal("Invalid reportInterval")

luckyseadog committed:
fix address for server

luckyseadog committed:
fix address

luckyseadog committed:
files

luckyseadog committed:
devide handlers in many files

luckyseadog committed:
goimports

luckyseadog committed:
fix typos

luckyseadog committed:
fix bug that caused situation that PollCount saved as Gauge

luckyseadog committed:
two values of unknown metrics

luckyseadog committed:
decided to save unknown metrics twice

luckyseadog committed:
file for saving metric types

luckyseadog committed:
add flags

luckyseadog committed:
minor fix

luckyseadog committed:
check address

luckyseadog committed:
concarrent metric syncronization

luckyseadog committed:
separate mutex

luckyseadog committed:
gzip support

luckyseadog committed:
change stroe interval parsing

luckyseadog committed:
moved the code from server/main to server/setup

luckyseadog committed:
add mechanism for signing transmitted data

luckyseadog committed:
hex encoding/decoding

luckyseadog committed:
ValueJSON handler returns hash

avhitrov commented on internal/handlers/handler_update_json.go:
```diff
@@ -0,0 +1,120 @@
+	}
+
+	for _, metric := range metricsCurrent {
+		if metric.MType == "gauge" {
```
В go для подобных конструкций принято использовать switch.
```
switch metric.MType {
case "gauge":
case "counter":
default:
}

avhitrov replied:
??

avhitrov commented on internal/handlers/handler_update_json.go:
```diff
@@ -0,0 +1,120 @@
+		_, err = w.Write(jsonData[1 : len(jsonData)-1])
+	} else {
+		_, err = w.Write(jsonData)
+	}
```
Это какая-то самодеятельность. Не видел в ТЗ указания ломать протокол передачи данных. На вход и на выход в любом случае должен передаваться список структур. Убрать! И обратное преобразование выше - тоже.

avhitrov commented on internal/handlers/handler_update_json.go:
```diff
@@ -0,0 +1,120 @@
+	if body[0] != byte('[') {
+		body = append([]byte{byte('[')}, body...)
+		body = append(body, byte(']'))
+	}
```
Оверинжиниринг. Если пришел не список, должна прилетать ошибка о неверном запросе.

avhitrov commented on internal/handlers/handler_value_json.go:
```diff
@@ -0,0 +1,98 @@
+	if body[0] != byte('[') {
+		body = append([]byte{byte('[')}, body...)
+		body = append(body, byte(']'))
+	}
```
То же, что и выше. Не надо ломать протокол общения сервисов.

avhitrov commented on cmd/server/main.go:
```diff
@@ -21,14 +58,47 @@ func main() {
+		r.Post("/", func(w http.ResponseWriter, r *http.Request) {
+			handlers.HandlerUpdateJSON(w, r, s)
+			select {
+			case <-fileSaveChan:
```
Способ, не гарантирующий синхронную запись, в первую очередь потому, что запрос может попасть на "перезарядку" генератора сигналов и уйти в default: В итоге по одному запросу сохранение выполнится, а закончившийся практически в одно время с ним другой запрос не приведет к сохранению. Лучше функцию SaveToFile дергать в горутинах, которые запускает PassSignal, а переданный ей канал (который она будет слушать) передать в in-memory storage и дергать его при условии, что флаг синхронной записи взведен.

avhitrov commented on internal/agent/agent_methods.go:
```diff
@@ -0,0 +1,108 @@
+		case <-a.cancel:
+			return
+		default:
+			<-ticker.C
```
```suggestion
		case <-ticker.C:
			
```

avhitrov commented on internal/agent/agent_methods.go:
```diff
@@ -0,0 +1,108 @@
+		case <-a.cancel:
+			return
+		default:
+			<-ticker.C
```
```suggestion
		case <-ticker.C:
```

avhitrov commented on internal/agent/agent_methods.go:
```diff
@@ -0,0 +1,108 @@
+			a.mu.Lock()
+			runtime.ReadMemStats(&a.metrics.MemStats)
+			a.metrics.RandomValue = metrics.Gauge(rand.Float64())
+			a.metrics.PollCount += 1
```
Из-за того, что счетчик после каждой успешной отправки данных не сбрасывается, значение на сервере растет нелинейно и не отображает вообще ничего. Можно убедиться, подергав в браузере страницу http://127.0.0.1:8080/value/counter/PollCount

avhitrov commented on internal/server/server_setup.go:
```diff
@@ -0,0 +1,80 @@
+	flag.StringVar(&addressFlag, "a", "127.0.0.1:8080", "address of server")
+	flag.StringVar(&storeIntervalStrFlag, "i", "300", "time to make new write in disk")
+	flag.StringVar(&storeFileFlag, "f", "/tmp/devops-metrics-db.json", "file in which we are saving metrics")
+	flag.StringVar(&restoreStrFlag, "r", "true", "if it is needed to load metrics from the past")
```
[BoolVar](https://pkg.go.dev/flag#BoolVar)

avhitrov commented on cmd/server/main.go:
```diff
@@ -12,6 +17,38 @@ import (
+
+	envVariables := server.SetUp()
+
+	dir := filepath.Dir(envVariables.StoreFile)
```
В эту переменную передается файл, в который необходимо складывать значения переданных параметров. 
1. Для чего-то создается файл metric_types.json, в который валится значение глобальной проектной переменной.
2. Файл блокируется до конца работы сервиса.

Нет ни смысла, ни отдельного указания в задании на создание этого файла.

avhitrov commented on internal/storage/metrics_syncronization.go:
```diff
@@ -0,0 +1,61 @@
+	"time"
+)
+
+func (s *MyStorage) SaveMetricsTypes(cancelChan chan struct{}, filepath string) {
```
Для чего все это?

avhitrov reviewed (changes requested):
Довольно много замечаний по коду. Не работает сохранение данных в файл. То есть при запуске сервера с параметрами
```
.\server.exe -a :8080 -f="database.json" -i=0
```
не создается `database.json`, зато для чего-то создается `metric_types.json`, в который со скоростью раз в миллисекунду записывается значение глобальной переменной.

luckyseadog committed:
Fix following errors:

Now PollCount will be equal to zero after PostStats
Remove saving metric types to file and instead of it made struct with gauge metrics and counter metrics

luckyseadog committed:
fix

luckyseadog committed:
Revert "fix"

This reverts commit 12eb5fd4c717ba8048580fa6fb31e4bf1a00196e.

luckyseadog committed:
fix

luckyseadog committed:
fixed magor problems

luckyseadog committed:
fix

luckyseadog committed:
fixed issue with sending []struct instead of struct

luckyseadog committed:
changed logic of saving data to file: storage pass signal in case of sync updates

luckyseadog committed:
fix

avhitrov commented on cmd/server/main.go:
```diff
@@ -21,14 +30,31 @@ func main() {
-	defer server.Close()
+	srv := server.NewServer(envVariables.Address, middlewares.GzipMiddleware(r))
+	srv.Run()
+	defer srv.Close()
```
defer в конце функции не имеет смысла, так как его функция и так выполнится последней.

avhitrov commented on internal/agent/agent_methods.go:
```diff
@@ -0,0 +1,115 @@
+)
+
+func (a *Agent) GetStats() {
+	ticker := time.NewTicker(a.ruler.pollInterval)
```
Тикер после использования необходимо обязательно освобождать путем вызова функции `Stop()`
```
defer ticker.Stop()
```
https://pkg.go.dev/time#Ticker

avhitrov commented on internal/agent/agent_methods.go:
```diff
@@ -0,0 +1,115 @@
+}
+
+func (a *Agent) PostStats() {
+	ticker := time.NewTicker(a.ruler.reportInterval)
```
https://pkg.go.dev/time#Ticker

avhitrov commented on internal/handlers/handler_update.go:
```diff
@@ -0,0 +1,84 @@
+	metricType, metric, metricValueString := splitPath[len(splitPath)-3],
+		metrics.Metric(splitPath[len(splitPath)-2]), splitPath[len(splitPath)-1]
+
+	if metricType == "gauge" {
```
В go для подобных конструкций принято использовать switch.
```
switch metric.MType {
case "gauge":
case "counter":
default:
}

avhitrov commented on internal/storage/storage.go:
```diff
@@ -1,67 +1,199 @@
+	return copyDataCounter
+}
+
+func (s *MyStorage) SaveToFile(filepath string) error {
```
В задании нет указания сохранять историю всех состояний. Поэтому достаточно держать в файле только последний слепок состояния. Для этого больше подойдет os.WriteFile (https://pkg.go.dev/os#WriteFile) и os.ReadFile, которые не надо даже обвешивать какими-либо сканнерами или ридерами.

avhitrov commented on internal/storage/storage.go:
```diff
@@ -1,67 +1,199 @@
+	return writer.Flush()
+}
+
+func (s *MyStorage) LoadFromFile(filepath string) error {
```
Если хранить только одно состояние, то функция сводится к os.ReadFile (https://pkg.go.dev/os#ReadFile) и анмаршалингу лежащего внутри json.

avhitrov reviewed (changes requested):
Сохранение информации в файл сделано не в соответствии с заданием (зачем-то сохраняется вся история, файл постоянно распухает). Осталось несколько не обработанных замечаний с прошлого ревью плюс добавил несколько новых, в основном, касающихся освобождения ресурсов. 

Реализация выбора метода сохранения сделана качественно, приятно посмотреть.

luckyseadog committed:
changes according review

avhitrov reviewed (approved):
Все замечания учтены. Добро пожаловать на следующий уровень!
